druid-shell/src/platform/mac/alert.rs

```diff
@@ -0,0 +1,120 @@
+        None => (), // The OS will default to NSAlertStyle::Warning
+    }
+
+    // The primary button is always first. (macOS alert buttons go from right to left)
+    let btn_primary = alert.addButtonWithTitle(make_nsstring(&request.options.primary.label));
+    // Set the primary button as the default which can be chosen via Enter.
+    btn_primary.setKeyEquivalent(make_nsstring("\r"));
+    btn_primary.setTag(BUTTON_ID_PRIMARY);
+
+    // If there's a cancel button, it is always second.
+    if let Some(cancel) = &request.options.cancel {
+        let btn_cancel = alert.addButtonWithTitle(make_nsstring(&cancel.label));
+        // Mark it as the cancel button by adding the Escape hotkey. This also activates Cmd+dot.
+        btn_cancel.setKeyEquivalent(make_nsstring("\u{1b}"));
+        btn_cancel.setTag(BUTTON_ID_CANCEL);
+    }
+
+    // Any alternative buttons are always last.
+    for (idx, alternative) in request.options.alternatives.iter().enumerate() {
+        let btn_alt = alert.addButtonWithTitle(make_nsstring(&alternative.label));
+        // Any hotkeys need to be explicit with custom buttons to avoid duplicates.
+        btn_alt.setKeyEquivalent(make_nsstring(""));
+        btn_alt.setTag(BUTTON_ID_ALT_START + idx as NSInteger);
+    }
+
+    if request.options.app_modal {
+        let tag = alert.runModal();
+        handle_result(view, &request, tag);
+    } else {
+        let window: id = msg_send![view, window];
+        let view_ptr = WeakPtr::new(view as *mut Object);
+
+        let alert_handler = block::ConcreteBlock::new(move |tag: NSInteger| {
+            let view = *view_ptr.load();
+            if !view.is_null() {
+                handle_result(&mut *view, &request, tag);
+            }
+        });
+        let alert_handler = alert_handler.copy(); // Force the block to live on the heap
+
+        alert.beginSheetModalForWindow_completionHandler(window, alert_handler);
+    }
+}
+
+unsafe fn handle_result(view: &mut Object, request: &AlertRequest, tag: NSInteger) {
+    let button = match tag {
+        BUTTON_ID_PRIMARY => Some(request.options.primary.clone()),
+        BUTTON_ID_CANCEL | NSModalResponseAbort | NSModalResponseStop | NSModalResponseCancel => {
+            None
+        }
+        id => {
+            if id >= BUTTON_ID_ALT_START
+                && id < BUTTON_ID_ALT_START + request.options.alternatives.len() as NSInteger
+            {
+                Some(request.options.alternatives[(id - BUTTON_ID_ALT_START) as usize].clone())
+            } else {
+                log::error!("Unexpected alert dialog result {}", id);
+                None
+            }
+        }
+    };
+    let view_state = {
+        let view_state: *mut c_void = *view.get_ivar("viewState");
+        &mut *(view_state as *mut ViewState)
+    };
+    (*view_state)
+        .handler
+        .alert_response(AlertResponse::new(request.token, button));
```

Comment:
From a casual look it feels like this is the only unsafe in this function? In that case I might either split it into a separate `unsafe` function or else make this function `safe` and just use an `unsafe` block here.

Edit: my thinking here is that we always want to make unsafe blocks as small as possible, because they demand much more scrutiny during review or even just reading the code; in particular future changes to this function have to be carefully reviewed for soundness.